ktor-server-rate-limiting/src/commonMain/kotlin/io/github/flaxoos/ktor/server/plugins/ratelimiter/implementations/Bucket.kt

```diff
@@ -0,0 +1,143 @@
+            volume - by.consideringLastUpdateTime(timeSinceLastUpdate).coerceAtLeast(0.0)
+        }
+    }
+
+    /**
+     * Increases the volume by the given amount if it won't lead to a volume exceeding the capacity
+     *
+     * @param shouldUpdateTime whether to update the last update time
+     * @param by the amount to increase the volume by, must be positive
+     */
+    internal suspend fun tryIncreaseVolume(
+        call: ApplicationCall,
+        by: Double,
+        shouldUpdateTime: Boolean = false,
+        consideringLastUpdateTime: suspend Double.(Long) -> Double = { this }
+    ): Double? {
+        by.assertNotNegative()
+        return increaseVolume(call, by, shouldUpdateTime, consideringLastUpdateTime)
+    }
+
+    /**
+     * Increases the volume by the given amount, coercing to at most the capacity
+     *
+     * @param shouldUpdateTime whether to update the last update time
+     * @param by the amount to increase the volume by, must be positive
+     */
+    internal suspend fun increaseVolume(
+        call: ApplicationCall,
+        by: Double,
+        shouldUpdateTime: Boolean = false,
+        consideringLastUpdateTime: suspend Double.(Long) -> Double = { this }
+    ): Double? {
+        by.assertNotNegative()
+        return tryUpdateVolume(call, shouldUpdateTime) { volume, timeSinceLastUpdate ->
+            volume + (consideringLastUpdateTime(by, timeSinceLastUpdate)).coerceAtMost(capacity.toDouble())
+        }
+    }
+
+    /**
+     * Tries to update the volume if the updated volume is within the capacity of the bucket
+     *
+     * @param update the function to update the volume, with the current volume and the time since the last update as
+     * the arguments
+     * @return the updated volume if it is within the capacity of the bucket,
+     * or null if the updated volume is not within the capacity
+     */
+    private suspend fun tryUpdateVolume(
+        call: ApplicationCall,
+        shouldUpdateTime: Boolean = false,
+        update: suspend (Double, Long) -> Double
+    ): Double? {
+        return currentVolumeMutex.withLock {
+            val now = clock()
+            val lastUpdateTime = _lastUpdateTime.value
+            val timeSinceLastUpdate = now - lastUpdateTime
+            log.trace {
+                "${call.id()}: Maybe updating volume at now: ${fromEpochMilliseconds(now)}, " +
+                    "lastUpdateTime: ${fromEpochMilliseconds(lastUpdateTime)}, " +
+                    "diff: $timeSinceLastUpdate"
+            }
+
+            val updatedVolume = update(currentVolume, timeSinceLastUpdate)
+            val diff = updatedVolume - currentVolume
+
+            if (
+                0 <= updatedVolume &&
+                updatedVolume <= capacity * callVolumeUnit.size
+            ) {
+                log.trace { "${call.id()}: Updating volume from $currentVolume to $updatedVolume" }
```

Comment:
Is there an option to disable of enable logging? Also, there are no guards around logging (if statements)

Reply:
there is inherently in how kotlin logging works, notice that the logged value is passed in a lambda, so it's only evaluated if logging happens:

```kotlin
public interface KLogger {

  /** Lazy add a log message if isTraceEnabled is true */
  public fun trace(message: () -> Any?): Unit =
    at(Level.TRACE) { this.message = message.toStringSafe() }
    ```

Reply:
Yes, that's how logging is generally implemented. You can enable / disable globally. 

BUT - you do not let the user control enabling or disabling logging for the Middleware specifically, no? Neither do you allow users to pass in a configured logger. Hence the question.

Reply:
this is all controlled in the logging configuration in the java/kotlin world:
```
<configuration>
    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
        <encoder>
            <pattern>%d{YYYY-MM-dd HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
        </encoder>
    </appender>
    <root level="info">
        <appender-ref ref="STDOUT"/>
    </root>
    <logger name="io.netty" level="INFO"/>
    <logger name="io.ktor" level="debug"/>
    <logger name="io.github.flaxoos.ktor.server.plugins.ratelimiter.implementations" level="debug"/>
</configuration>
```